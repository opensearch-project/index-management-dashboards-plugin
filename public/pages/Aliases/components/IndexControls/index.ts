/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import IndexControls from "./IndexControls";

export * from "./IndexControls";
export default IndexControls;
