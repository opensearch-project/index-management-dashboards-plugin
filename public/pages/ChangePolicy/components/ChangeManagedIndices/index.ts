/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import ChangeManagedIndices from "./ChangeManagedIndices";

export default ChangeManagedIndices;
