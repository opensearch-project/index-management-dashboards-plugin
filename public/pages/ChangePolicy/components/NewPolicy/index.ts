/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import NewPolicy from "./NewPolicy";

export default NewPolicy;
