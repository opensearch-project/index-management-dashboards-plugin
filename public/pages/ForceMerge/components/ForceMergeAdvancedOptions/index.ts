/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import ForceMergeAdvancedOptions from "./ForceMergeAdvancedOptions";

export default ForceMergeAdvancedOptions;
export * from "./ForceMergeAdvancedOptions";
