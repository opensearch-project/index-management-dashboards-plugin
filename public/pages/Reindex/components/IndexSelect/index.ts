/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import IndexSelect from "./IndexSelect";

export default IndexSelect;
