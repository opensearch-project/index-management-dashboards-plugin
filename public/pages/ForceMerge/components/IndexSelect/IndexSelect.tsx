/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiComboBoxOptionOption } from "@elastic/eui";
import { _EuiComboBoxProps } from "@elastic/eui/src/components/combo_box/combo_box";
import { CoreStart } from "opensearch-dashboards/public";
import { debounce } from "lodash";
import React, { useContext, useEffect, useMemo, useRef, useState } from "react";
import ComboBoxWithoutWarning from "../../../../components/ComboBoxWithoutWarning";
import { CoreServicesContext } from "../../../../components/core_services";
import { IndexSelectItem } from "../../models/interfaces";
import { filterByMinimatch } from "../../../../../utils/helper";
import { SYSTEM_ALIAS, SYSTEM_INDEX } from "../../../../../utils/constants";

interface IndexSelectProps extends Pick<_EuiComboBoxProps<IndexSelectItem>, "data-test-subj" | "placeholder"> {
  getIndexOptions: (searchValue: string) => Promise<Array<EuiComboBoxOptionOption<IndexSelectItem>>>;
  onChange: (options: string[]) => void;
  singleSelect: boolean;
  value?: string[];
  excludeSystemIndex?: boolean;
}

export default function IndexSelect(props: IndexSelectProps) {
  const [indexOptions, setIndexOptions] = useState([] as Array<EuiComboBoxOptionOption<IndexSelectItem>>);
  const coreServices = useContext(CoreServicesContext) as CoreStart;
  const destroyRef = useRef(false);

  const searchIndex = (searchValue?: string) => {
    props
      .getIndexOptions(searchValue ? searchValue : "")
      .then((options) => {
        if (props.excludeSystemIndex) filterSystemIndices(options);
        if (!destroyRef.current) {
          setIndexOptions(options);
        }
      })
      .catch((err) => {
        coreServices.notifications.toasts.addDanger(`fetch indices error ${err}`);
      });
  };

  useEffect(
    () => {
      searchIndex();
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [props.getIndexOptions, props.excludeSystemIndex]
  );

  useEffect(() => {
    return () => {
      destroyRef.current = true;
    };
  }, []);

  const onSearchChangeRef = useRef(
    debounce(
      (searchValue: string) => {
        searchIndex(searchValue);
      },
      200,
      {
        leading: true,
      }
    )
  );

  const filterSystemIndices = (list: Array<EuiComboBoxOptionOption<IndexSelectItem>>) => {
    list.map((it) => {
      it.options = it.options?.filter((item) => !filterByMinimatch(item.label, SYSTEM_ALIAS));
      it.options = it.options?.filter((item) => !filterByMinimatch(item.label, SYSTEM_INDEX));
    });
  };

  const flattenedOptions = useMemo(
    () =>
      indexOptions.reduce(
        (total, current) => [...total, ...(current.options || [])],
        [] as Array<EuiComboBoxOptionOption<IndexSelectItem>>
      ),
    [indexOptions]
  );

  const finalSelectedOptions: Array<EuiComboBoxOptionOption<IndexSelectItem>> =
    props.value
      ?.map((item) => flattenedOptions.find((option) => option.label === item) as EuiComboBoxOptionOption<IndexSelectItem>)
      .filter((item) => item) || [];

  return (
    <div>
      <ComboBoxWithoutWarning
        data-test-subj={props["data-test-subj"]}
        placeholder={props.placeholder}
        options={indexOptions}
        async
        selectedOptions={finalSelectedOptions}
        onChange={(selectedOptions) => props.onChange(selectedOptions.map((item) => item.label))}
        onSearchChange={onSearchChangeRef.current}
        isClearable={true}
        singleSelection={props.singleSelect ? { asPlainText: true } : false}
      />
    </div>
  );
}
