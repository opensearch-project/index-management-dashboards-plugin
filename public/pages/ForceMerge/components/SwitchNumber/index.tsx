/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import React, { useState } from "react";
import { EuiRadioGroup, EuiSpacer } from "@elastic/eui";
import CustomFormRow from "../../../../components/CustomFormRow";
import { AllBuiltInComponents } from "../../../../components/FormGenerator";

interface SwitchNumberProps {
  value?: number;
  onChange: (val: SwitchNumberProps["value"]) => void;
}

export default function SwitchNumber(props: SwitchNumberProps) {
  const [id, setId] = useState(props.value && props.value > 0 ? "1" : "0");
  return (
    <>
      <EuiRadioGroup
        options={[
          {
            id: "0",
            label: "Automatic",
          },
          {
            id: "1",
            label: "Manually set number of segments",
          },
        ]}
        idSelected={id}
        onChange={(id) => {
          setId(id);
          if (id === "0") {
            props.onChange(undefined);
          } else {
            props.onChange(1);
          }
        }}
      />
      {id === "1" ? (
        <>
          <EuiSpacer size="s" />
          <CustomFormRow
            label="Number of segments"
            helpText="Specify the number of segments to merge to. To fully merge indexes, set the number of segments to 1."
          >
            <AllBuiltInComponents.Number {...props} placeholder="Specify a number" />
          </CustomFormRow>
        </>
      ) : null}
    </>
  );
}
