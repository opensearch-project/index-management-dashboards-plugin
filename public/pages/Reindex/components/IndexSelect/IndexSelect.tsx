/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiComboBoxOptionOption } from "@elastic/eui";
import { _EuiComboBoxProps } from "@elastic/eui/src/components/combo_box/combo_box";
import { CoreStart } from "opensearch-dashboards/public";
import React, { useContext, useEffect, useState } from "react";
import ComboBoxWithoutWarning from "../../../../components/ComboBoxWithoutWarning";
import { CoreServicesContext } from "../../../../components/core_services";
import { IndexSelectItem } from "../../models/interfaces";
import { filterOverlaps } from "../../utils/helper";
import { filterByMinimatch } from "../../../../../utils/helper";
import { SYSTEM_ALIAS, SYSTEM_INDEX } from "../../../../../utils/constants";

interface IndexSelectProps extends Pick<_EuiComboBoxProps<IndexSelectItem>, "data-test-subj" | "placeholder"> {
  getIndexOptions: (searchValue: string, excludeDataStreamIndex?: boolean) => Promise<Array<EuiComboBoxOptionOption<IndexSelectItem>>>;
  onSelectedOptions: (options: Array<EuiComboBoxOptionOption<IndexSelectItem>>) => void;
  singleSelect: boolean;
  selectedOption: Array<EuiComboBoxOptionOption<IndexSelectItem>>;
  excludeDataStreamIndex?: boolean;
  excludeList?: Array<EuiComboBoxOptionOption<IndexSelectItem>>;
  excludeSystemIndex?: boolean;
}

export default function IndexSelect(props: IndexSelectProps) {
  const [indexOptions, setIndexOptions] = useState([] as Array<EuiComboBoxOptionOption<IndexSelectItem>>);
  const coreServices = useContext(CoreServicesContext) as CoreStart;

  const searchIndex = (searchValue?: string) => {
    props
      .getIndexOptions(searchValue ? searchValue : "", props.excludeDataStreamIndex)
      .then((options) => {
        if (props.excludeSystemIndex) filterSystemIndices(options);
        setIndexOptions(filterOverlaps(options, props.excludeList));
      })
      .catch((err) => {
        coreServices.notifications.toasts.addDanger(`fetch indices error ${err}`);
      });
  };

  useEffect(
    () => {
      searchIndex();
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [props.getIndexOptions, props.excludeList, props.excludeDataStreamIndex, props.excludeSystemIndex]
  );

  const onSearchChange = (searchValue: string) => {
    searchIndex(searchValue);
  };

  const filterSystemIndices = (list: Array<EuiComboBoxOptionOption<IndexSelectItem>>) => {
    list.map((it) => {
      it.options = it.options?.filter((item) => !filterByMinimatch(item.label, SYSTEM_ALIAS));
      it.options = it.options?.filter((item) => !filterByMinimatch(item.label, SYSTEM_INDEX));
    });
  };

  return (
    <div>
      <ComboBoxWithoutWarning
        data-test-subj={props["data-test-subj"]}
        placeholder={props.placeholder}
        options={indexOptions}
        async
        selectedOptions={props.selectedOption}
        onChange={props.onSelectedOptions}
        onSearchChange={onSearchChange}
        isClearable={true}
        singleSelection={props.singleSelect ? { asPlainText: true } : false}
      />
    </div>
  );
}
