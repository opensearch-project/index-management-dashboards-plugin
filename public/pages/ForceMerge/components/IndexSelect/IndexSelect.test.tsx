/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { render, waitFor } from "@testing-library/react";
import React from "react";
import { act } from "react-dom/test-utils";
import { coreServicesMock } from "../../../../../test/mocks";
import { CoreServicesContext } from "../../../../components/core_services";
import IndexSelect from "./IndexSelect";

describe("<IndexSelect /> spec", () => {
  it("renders the component", async () => {
    const component = render(
      <CoreServicesContext.Provider value={coreServicesMock}>
        <IndexSelect
          getIndexOptions={(str) => Promise.resolve([{ label: "sourceIndex" }])}
          onSelectedOptions={(options) => {}}
          selectedOption={[{ label: "sourceIndex" }]}
          singleSelect={true}
        />
      </CoreServicesContext.Provider>
    );
    await waitFor(() => {});
    expect(component).toMatchSnapshot();
  });

  it("renders the component with error", async () => {
    const getIndexOptionsFn = jest.fn().mockRejectedValue("service not available");
    act(() => {
      render(
        <CoreServicesContext.Provider value={coreServicesMock}>
          <IndexSelect
            getIndexOptions={getIndexOptionsFn}
            onSelectedOptions={(options) => {}}
            selectedOption={[{ label: "sourceIndex" }]}
            singleSelect={true}
          />
        </CoreServicesContext.Provider>
      );
    });
    await waitFor(() => {});
    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledTimes(1);
  });

  it("search async", async () => {
    const getIndexOptionsFn = jest.fn().mockResolvedValue([{ label: "test-index" }]);
    act(() => {
      render(
        <CoreServicesContext.Provider value={coreServicesMock}>
          <IndexSelect
            getIndexOptions={getIndexOptionsFn}
            onSelectedOptions={(options) => {}}
            selectedOption={[{ label: "sourceIndex" }]}
            singleSelect={false}
          />
        </CoreServicesContext.Provider>
      );
    });
    await waitFor(() => {});
    expect(getIndexOptionsFn).toHaveBeenCalledTimes(1);
  });
});
