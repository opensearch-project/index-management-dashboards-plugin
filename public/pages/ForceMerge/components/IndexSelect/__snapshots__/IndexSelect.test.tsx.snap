// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IndexSelect /> spec renders the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          class="euiComboBox"
          role="combobox"
        >
          <div
            class="euiFormControlLayout"
          >
            <div
              class="euiFormControlLayout__childrenWrapper"
            >
              <div
                class="euiComboBox__inputWrap euiComboBox__inputWrap--noWrap euiComboBox__inputWrap-isClearable"
                data-test-subj="comboBoxInput"
                tabindex="-1"
              >
                <div
                  class="euiComboBox__input"
                  style="font-size: 14px; display: inline-block;"
                >
                  <input
                    aria-controls=""
                    data-test-subj="comboBoxSearchInput"
                    role="textbox"
                    style="box-sizing: content-box; width: 2px;"
                    value=""
                  />
                  <div
                    style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                  />
                </div>
              </div>
              <div
                class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
              >
                <button
                  aria-label="Open list of options"
                  class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                  data-test-subj="comboBoxToggleListButton"
                  type="button"
                >
                  EuiIconMock
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        class="euiComboBox"
        role="combobox"
      >
        <div
          class="euiFormControlLayout"
        >
          <div
            class="euiFormControlLayout__childrenWrapper"
          >
            <div
              class="euiComboBox__inputWrap euiComboBox__inputWrap--noWrap euiComboBox__inputWrap-isClearable"
              data-test-subj="comboBoxInput"
              tabindex="-1"
            >
              <div
                class="euiComboBox__input"
                style="font-size: 14px; display: inline-block;"
              >
                <input
                  aria-controls=""
                  data-test-subj="comboBoxSearchInput"
                  role="textbox"
                  style="box-sizing: content-box; width: 2px;"
                  value=""
                />
                <div
                  style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                />
              </div>
            </div>
            <div
              class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
            >
              <button
                aria-label="Open list of options"
                class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                data-test-subj="comboBoxToggleListButton"
                type="button"
              >
                EuiIconMock
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
